lib/peer_manager/modules/peer_data_storage.c

```diff
@@ -273,224 +244,224 @@ static void fds_evt_handler(fds_evt_t const *const p_fds_evt)
+		break;
+	case BM_ZMS_EVT_DELETE:
+		pds_evt.params.peer_data_update_succeeded.data_id = data_id;
+		pds_evt.params.peer_data_update_succeeded.action = PM_PEER_DATA_OP_DELETE;
+		pds_evt.params.peer_data_update_succeeded.token = p_evt->id;
+
+		if (p_evt->result == 0) {
+			pds_evt.evt_id = PM_EVT_PEER_DATA_UPDATE_SUCCEEDED;
+			pds_evt.params.peer_data_update_succeeded.flash_changed = true;
+		} else {
+			pds_evt.evt_id = PM_EVT_PEER_DATA_UPDATE_FAILED;
+			pds_evt.params.peer_data_update_failed.error = p_evt->result;
+		}
+
+		pds_evt_send(&pds_evt);
+		break;
+	default:
+		/* No action. */
+		break;
 	}
 
-	peer_id_init();
-	peer_ids_load();
+	if (m_peer_delete_deferred) {
+		peer_data_delete_process();
+	}
+}
 
-	m_module_initialized = true;
+static void wait_for_init(void)
+{
+	while (!fs.init_flags.initialized) {
+#if defined(CONFIG_SOFTDEVICE)
+		/* Wait for an event. */
+		__WFE();
+
+		/* Clear Event Register */
+		__SEV();
+		__WFE();
+#endif
+	}
+}
 
-	return NRF_SUCCESS;
+void pds_peer_data_iterate_prepare(pm_peer_id_t *p_peer_id_iter)
+{
+	*p_peer_id_iter = 0;
 }
 
-uint32_t pds_peer_data_read(pm_peer_id_t peer_id, pm_peer_data_id_t data_id,
-			      pm_peer_data_t *const p_data, uint32_t const *const p_buf_len)
+bool pds_peer_data_iterate(pm_peer_data_id_t data_id, pm_peer_id_t *const p_peer_id,
+			   pm_peer_data_flash_t *const p_data, pm_peer_id_t *p_peer_id_iter)
 {
-	uint32_t ret;
-	fds_record_desc_t rec_desc;
-	fds_flash_record_t rec_flash;
+	ssize_t ret;
+	uint8_t temp_buf[PM_PEER_DATA_MAX_SIZE] = { 0 };
 
-	NRF_PM_DEBUG_CHECK(m_module_initialized);
-	NRF_PM_DEBUG_CHECK(p_data != NULL);
+	if (*p_peer_id_iter >= PM_PEER_ID_N_AVAILABLE_IDS) {
+		return false;
+	}
 
-	VERIFY_PEER_ID_IN_RANGE(peer_id);
-	VERIFY_PEER_DATA_ID_IN_RANGE(data_id);
+	/* Exits the loop when `ret > 0` (it found data), it reached the end of the available peers,
+	 * or the read had a catastrophical failure.
+	 */
+	do {
+		uint32_t entry_id = peer_id_peer_data_id_to_entry_id(*p_peer_id_iter, data_id);
+
+		ret = bm_zms_read(&fs, entry_id, temp_buf, PM_PEER_DATA_MAX_SIZE);
+		if (ret < 0 && ret != -ENOENT) {
+			LOG_ERR("Could not read data from NVM. bm_zms_read() returned %d. "
+				"peer_id: %d",
+				ret, *p_peer_id_iter);
+			return false;
+		}
 
-	ret = peer_data_find(peer_id, data_id, &rec_desc);
+		(*p_peer_id_iter)++;
+	} while ((ret == -ENOENT) && (*p_peer_id_iter < PM_PEER_ID_N_AVAILABLE_IDS));
 
-	if (ret != NRF_SUCCESS) {
-		return NRF_ERROR_NOT_FOUND;
+	if ((ret == -ENOENT) && (*p_peer_id_iter == PM_PEER_ID_N_AVAILABLE_IDS)) {
+		return false;
 	}
 
-	/* Shouldn't fail, unless the record was deleted in the meanwhile or the CRC check has
-	 * failed.
+	/* We found a suitable Peer ID. */
+
+	*p_peer_id = *p_peer_id_iter;
+
+	/* `ret` is equal the exact amount of data contained in the entry, so copy that amount
+	 * safely.
 	 */
-	ret = fds_record_open(&rec_desc, &rec_flash);
+	memcpy((void *)p_data->p_all_data, temp_buf, ret);
```

Comment:
Would it be better to pass `p_data->p_all_data` directly to `bm_zms_read`? It would avoid some stack usage and allow different max sizes.

In either case, `p_data->p_all_data` should maybe be null checked.

Reply:
If you add a `p_buf_len` like I suggested below, would this still be a problem?